risc0/zkvm/sdk/rust/guest/src/sha.rs

```diff
@@ -12,132 +12,178 @@
-        let trailer: *mut usize = ptr.add(total - WORD_SIZE).cast();
-        trailer.write_volatile(bits.to_be());
-
-        // Set up the next descriptor.
-        let desc = get_cur_desc();
-        desc.write_volatile(SHADescriptor {
-            type_count: total / 64,
-            idx: 0,
-            source: ptr as usize,
-            digest: result as usize,
-        });
-
-        // Write the descriptor to the oracle for processing.
-        GPIO_SHA.write_volatile(desc);
-
-        // Jump to the next descriptor.
-        CUR_DESC += 1;
+        let ptr = slice.as_mut_ptr();
+        let u8ptr = ptr as *mut u8;
+        assert!(len_bytes < slice.len() * WORD_SIZE);
+        u8ptr.add(len_bytes).write_volatile(END_MARKER);
     }
+
+    // Last word is size in bits.
+    let len_bits = len_bytes * 8;
+    slice[slice.len() - 1] = (len_bits as u32).to_be();
 }
 
-pub(crate) fn digest_commit_into(len: usize, slice: &mut [u32], result: *mut usize) {
-    let total = padded_size(len);
-    let len_words = len / WORD_SIZE;
-    slice[len_words] = 0x00000080;
-    for i in len_words + 1..total / WORD_SIZE - 1 {
-        slice[i] = 0;
-    }
-    finalize_into(len, total, slice.as_mut_ptr().cast(), result);
+// Digests the data in a SHA-ready buffer.  Consumes the given buffer
+// since it needs to modify it to add the trailer.
+fn digest_buf<T: ShaBuf>(mut data: T) -> &'static [u32; DIGEST_WORDS] {
+    let len_bytes = data.len() * WORD_SIZE;
+    add_trailer(&mut data, len_bytes);
+    raw_digest(data.as_slice())
 }
 
-pub fn digest<T>(data: T) -> Box<Digest> {
-    let mut sha = SHA256::with_capacity(mem::size_of::<T>());
-    sha.update(&data);
-    sha.finalize()
+// Computes the SHA256 digest of a serialized object.
+pub fn digest<T: Serialize>(val: &T) -> &'static [u32; DIGEST_WORDS] {
+    // If the object to be serialized is a plain old structure in memory, this
+    // should be a good guess for the allocation needed.
+    let cap = compute_capacity_needed(mem::size_of_val(val));
+    let buf = to_vec_with_capacity(val, cap).unwrap();
+    digest_buf(buf)
 }
 
-pub fn digest_slice<T>(data: &[T]) -> Box<Digest> {
-    let mut sha = SHA256::with_capacity(data.len());
-    sha.update_slice(data);
-    sha.finalize()
+// Makes a digest for a slice of u8s.  We have no guarantees on
+// alignment so we have to copy the whole thing to a new buffer.
+pub fn digest_u8(data: &[u8]) -> &'static [u32; DIGEST_WORDS] {
+    let len_bytes = data.len();
+    let cap = compute_capacity_needed(len_bytes);
+    let mut data_u32 = Vec::<u32>::with_capacity(cap);
+    unsafe {
+        (data_u32.as_mut_ptr() as *mut u8).copy_from_nonoverlapping(data.as_ptr(), len_bytes);
+        data_u32.set_len(align_up(len_bytes, WORD_SIZE) / WORD_SIZE);
+    }
+    add_trailer(&mut data_u32, len_bytes);
+    raw_digest(data_u32.as_slice())
 }
 
+// Do some magic to finialize something relating to the guest environment.
+// TODO(nils): Document this better.
```

Comment:
What this is intending to do is set a marker so that the VM knows when the last SHA descriptor is reached. We need to write to this memory location at least once so that it's not uninitialized, and the value of 0 is the marker that the VM looks for.